Applications/Spire/Source/TimeAndSales/TimeAndSalesWindow.cpp

```diff
@@ -1,40 +1,183 @@
+  update_style(*m_body, [] (auto& style) {
     style.get(Any()).set(BackgroundColor(QColor(0xFFFFFF)));
   });
-  set_body(box);
+  set_body(m_body);
   resize(security_view->sizeHint().width(), scale_height(361));
+  m_table_model->connect_begin_loading_signal(
+    std::bind_front(&TimeAndSalesWindow::on_begin_loading, this));
+  m_table_model->connect_end_loading_signal(
+    std::bind_front(&TimeAndSalesWindow::on_end_loading, this));
+  m_table_view->get_table()->connect_operation_signal(
+    std::bind_front(&TimeAndSalesWindow::on_table_operation, this));
+}
+
+bool TimeAndSalesWindow::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == &m_table_view->get_header()) {
+    if(event->type() == QEvent::MouseButtonPress) {
+      auto& mouse_event = *static_cast<QMouseEvent*>(event);
+      if(mouse_event.button() == Qt::RightButton) {
+        m_table_header_menu->window()->move(mouse_event.globalPos());
+        m_table_header_menu->window()->show();
+      }
+    }
+  }
+  return Window::eventFilter(watched, event);
+}
+
+void TimeAndSalesWindow::make_table_header_menu() {
+  if(m_table_header_menu) {
+    return;
+  }
+  m_table_header_menu = new ContextMenu(*this);
+  auto header_items = m_table_view->get_header().get_items();
+  for(auto i = 0; i < header_items->get_size() - 1; ++i) {
+    auto model =
+      m_table_header_menu->add_check_box(header_items->get(i).m_name);
+    model->set(m_table_view->get_header().get_item(i)->isVisible());
+    model->connect_update_signal(std::bind_front(
+      &TimeAndSalesWindow::on_header_item_check, this, i));
+  }
+}
+
+void TimeAndSalesWindow::on_begin_loading() {
+  if(m_transition_view->get_status() == TransitionView::Status::NONE) {
+    m_transition_view->set_status(TransitionView::Status::LOADING);
+  }
+}
+
+void TimeAndSalesWindow::on_end_loading() {
+  m_transition_view->set_status(TransitionView::Status::READY);
+  make_table_header_menu();
 }
 
 void TimeAndSalesWindow::on_current(const Security& security) {
   setWindowTitle(to_text(security) + " " + QString(0x2013) + " " + TITLE_NAME);
-  m_transition_view->set_status(TransitionView::Status::LOADING);
+  m_transition_view->set_status(TransitionView::Status::NONE);
+  m_table_model->set_model(m_model_builder(security));
+  auto& header = m_table_view->get_header();
+  auto& properties = m_factory->make()->get_current()->get();
+  m_table_model->load_history(
+    (m_body->height() - header.height()) / get_height(properties.get_font()));
+}
+
+void TimeAndSalesWindow::on_header_item_check(int column, bool checked) {
+  m_table_view->get_header().get_item(column)->setVisible(checked);
+}
+
+void TimeAndSalesWindow::on_table_operation(
+    const TableModel::Operation& operation) {
+  visit(operation,
+    [&] (const TableModel::AddOperation& operation) {
+      auto indicator = m_table_model->get_bbo_indicator(operation.m_index);
+      for(auto i = 0; i < m_table_model->get_column_size(); ++i) {
+        auto& cell =
```

Comment:
This still has the same problem as before and smears out object construction. Keep logically similar code together instead of spreading it over the codebase. There is a specific place where cells are constructed and initialized and any work involving those cells should be kept there.